test/e2e/cluster_upgrade.go

```diff
@@ -35,53 +36,176 @@ import (
 	. "github.com/onsi/gomega"
 )
 
-// version applies to upgrades; kube-push always pushes local binaries.
-const version = "latest_ci"
+const (
+	// version applies to upgrades; kube-push always pushes local binaries.
+	version    = "latest_ci"
+	versionURL = "https://storage.googleapis.com/kubernetes-release/%s/%s.txt"
+)
+
+// realVersion turns a version constant--one accepted by cluster/gce/upgrade.sh--
+// into a deployable version string.
+//
+// NOTE: KEEP THIS LIST UP-TO-DATE WITH THE CODE BELOW.
+// The version strings supported are:
+// - "latest_stable"   (returns a string like "0.18.2")
+// - "latest_release"  (returns a string like "0.19.1")
+// - "latest_ci"       (returns a string like "0.19.1-669-gabac8c8")
+func realVersion(s string) (string, error) {
+	bucket, file := "", ""
+	switch s {
+	// NOTE: IF YOU CHANGE THE FOLLOWING LIST, ALSO UPDATE cluster/gce/upgrade.sh
+	case "latest_stable":
+		bucket, file = "release", "stable"
+	case "latest_release":
+		bucket, file = "release", "latest"
+	case "latest_ci":
+		bucket, file = "ci", "latest"
+	default:
+		return "", fmt.Errorf("version %s is not supported", s)
+	}
+
+	url := fmt.Sprintf(versionURL, bucket, file)
+	var v string
+	Logf("Fetching version from %s", url)
+	c := &http.Client{Timeout: 2 * time.Second}
+	if err := wait.Poll(poll, singleCallTimeout, func() (bool, error) {
+		r, err := c.Get(url)
+		if err != nil {
+			Logf("Error reaching %s: %v", url, err)
+			return false, nil
+		}
+		if r.StatusCode != http.StatusOK {
+			Logf("Bad response; status: %d, response: %v", r.StatusCode, r)
+			return false, nil
+		}
+		defer r.Body.Close()
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			Logf("Could not read response body: %v", err)
+			return false, nil
+		}
+		v = strings.TrimSpace(string(b))
+		return true, nil
+	}); err != nil {
+		return "", fmt.Errorf("failed to fetch real version from %s", url)
+	}
+	// Versions start with "v", so remove that.
+	return v[1:], nil
+}
 
 // The following upgrade functions are passed into the framework below and used
 // to do the actual upgrades.
+var masterUpgrade = func(v string) error {
+	var err error
```

Comment:
I was originally doing it to match `nodeUpgrade` (which does stuff after the call to validate), but I guess might as well just shortcut now given there's nothing after it. Done. (Not yet pushed.)